internal/oidc/keys.go

```diff
@@ -0,0 +1,195 @@
+	manager = NewKeyManager()
+
+	_, _, err = manager.AddActiveKeyData(configuration.IssuerPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return manager, nil
+}
+
+// NewKeyManager creates a new empty KeyManager.
+func NewKeyManager() (manager *KeyManager) {
+	manager = new(KeyManager)
+	manager.keys = map[string]*rsa.PrivateKey{}
+	manager.keySet = new(jose.JSONWebKeySet)
+
+	return manager
+}
+
+// Strategy returns the RS256JWTStrategy.
+func (m KeyManager) Strategy() (strategy *RS256JWTStrategy) {
+	return m.strategy
+}
+
+// GetKeySet returns the joseJSONWebKeySet containing the rsa.PublicKey types.
+func (m KeyManager) GetKeySet() (keySet *jose.JSONWebKeySet) {
+	return m.keySet
+}
+
+// GetActiveWebKey obtains the currently active jose.JSONWebKey.
+func (m KeyManager) GetActiveWebKey() (webKey *jose.JSONWebKey, err error) {
+	webKeys := m.keySet.Key(m.activeKeyID)
+	if len(webKeys) == 1 {
+		return &webKeys[0], nil
+	}
+
+	if len(webKeys) == 0 {
+		return nil, errors.New("could not find a key with the active key id")
+	}
+
+	return &webKeys[0], errors.New("multiple keys with the same key id")
+}
+
+// GetActiveKeyID returns the key id of the currently active key.
+func (m KeyManager) GetActiveKeyID() (keyID string) {
+	return m.activeKeyID
+}
+
+// GetActiveKey returns the rsa.PublicKey of the currently active key.
+func (m KeyManager) GetActiveKey() (key *rsa.PublicKey, err error) {
+	if key, ok := m.keys[m.activeKeyID]; ok {
+		return &key.PublicKey, nil
+	}
+
+	return nil, errors.New("failed to retrieve active public key")
+}
+
+// GetActivePrivateKey returns the rsa.PrivateKey of the currently active key.
+func (m KeyManager) GetActivePrivateKey() (key *rsa.PrivateKey, err error) {
+	if key, ok := m.keys[m.activeKeyID]; ok {
+		return key, nil
+	}
+
+	return nil, errors.New("failed to retrieve active private key")
+}
+
+// AddActiveKeyData adds a rsa.PublicKey given the key in the PEM string format, then sets it to the active key.
+func (m *KeyManager) AddActiveKeyData(data string) (key *rsa.PrivateKey, webKey *jose.JSONWebKey, err error) {
+	key, err = utils.ParseRsaPrivateKeyFromPemStr(data)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	webKey, err = m.AddActiveKey(key)
+
+	return key, webKey, err
+}
+
+// AddActiveKey adds a rsa.PublicKey, then sets it to the active key.
+func (m *KeyManager) AddActiveKey(key *rsa.PrivateKey) (webKey *jose.JSONWebKey, err error) {
+	wk := jose.JSONWebKey{
+		Key:       &key.PublicKey,
+		Algorithm: "RS256",
+		Use:       "sig",
+	}
+
+	keyID, err := wk.Thumbprint(crypto.SHA1)
+	if err != nil {
+		return nil, err
+	}
+
+	strKeyID := strings.ToLower(fmt.Sprintf("%x", keyID)[0:6])
+
+	if _, ok := m.keys[strKeyID]; ok {
+		return nil, fmt.Errorf("key id %s already exists", strKeyID)
+	}
+
+	// TODO: Add Mutex here when implementing key rotation.
+	wk.KeyID = strKeyID
+	m.keySet.Keys = append(m.keySet.Keys, wk)
+	m.keys[strKeyID] = key
+	m.activeKeyID = strKeyID
+
+	if m.strategy == nil {
```

Comment:
I would avoid the lazy assignment if possible and rather assign the strategy when the manager is created unless I missed something.

Reply:
This was done because it should be set when we assign the active key. However the nil check is not useful I don't think.

Reply:
Can you take a look at this adjustment? I think it's adequate.

Reply:
I'm fine with the current implementation but I'm wondering what prevents us to store the strategy in each key instead of at the key manager level? Maybe an improvement for a future review.